src/mug/ssl.gleam

```diff
@@ -0,0 +1,334 @@
+import gleam/bytes_builder.{type BytesBuilder}
+import gleam/dynamic.{type Dynamic}
+import gleam/erlang/atom
+import gleam/erlang/charlist.{type Charlist}
+import gleam/erlang/process
+import gleam/io
+import gleam/result
+import mug.{type ConnectionOptions, type Error}
+
+pub type Socket
+
+type DoNotLeak
+
+pub type SSLConnectionOptions {
+  SSLConnectionOptions(
+    host: String,
+    port: Int,
+    timeout: Int,
+    cacerts: CaCertificates,
+  )
+}
+
+/// The CA certificates to use
+pub type CaCertificates {
+  /// Do not verify certificates.
+  NoVerification
+  /// Use these der-encoded certificates as CA certificates. This will make mug not use
+  /// the system's CA certificates, but only these ones. To use the system's certificates
+  /// and custom ones, use the `WithSystemCertificatesAnd` variant instead.
+  CustomDerCertificates(cacerts: List(BitArray))
+  /// Path to a pem-encoded file which contains CA certificates. This will make mug not use
+  /// the system's CA certificates, but only these ones. To use the system's certificates
+  /// and custom ones, use the `WithSystemCertificatesAnd` variant instead.
+  CustomPemCertificates(cacertfile: String)
+  /// Use the system's CA certificates, as provided by erlang's `cacerts_get/0` function
+  /// from the `public_key` module ([docs](https://www.erlang.org/doc/apps/public_key/public_key#cacerts_get/0)).
+  SystemCertificates
+  /// Use the system's certificates, along with `cacerts` more, which is a list of
+  /// der-encoded CA certificates as BitArrays. A pem-file cannot be used with this option
+  /// because der-encoded certs have a higher priority over pem-encoded cert files, and the
+  /// `cacerts_get/0` function returns der-encoded certs.
+  WithSystemCertificatesAnd(cacerts: List(BitArray))
```

Comment:
This name seems not in keeping with the others, and uses an usual format. What else could we call this?

Reply:
Maybe I can drop the `With`? I'm all out of ideas :shrug:

Reply:
Looking at this again I think the whole type may need another look. It says it's cacerts, but it has a variant called `NoVerification`, which isn't a cacert.

Perhaps something like this:

```gleam
pub type Security {
  VerifyTls(cacerts: List(CaCerts))
  DangerouslySkipTlsVerification
}

pub type CaCerts {
  SystemCaCertificates
  CustomDerCaCertificates(cacerts: List(BitArray))
  CustomPemCaCertificates(cacertfile: String)
}
```

Then folks can provide whatever cacerts they wish

Reply:
This is much better!

Reply:
Using an Option would probably be better than using `Security`, which imho is a poorly named type (since security can mean anything, not just something to do with certificates). I can add some convenience methods like `no_verification`, which sets the certificates to `None`, and `use_system_certificates`, which sets the certificates to the system CA certs.

Reply:
It must not be an option, it must use the descriptive names. I am open to a different name for the type if you have a better one.

Reply:
Please use the name I used above that makes it extremely clear when you are disabling tls verification. It is vital that someone unfamiliar with this area understands what they are doing.

Reply:
I've changed the type name and field to verification, but I feel that NoVerification should stick, since that's what many other languages/libraries use. I've left some thoughts in the later inline comment you've put.

I'll mark this resolved since the code is outdated. Let's continue discussion in the inline comment occuring later on.